src/Aspire.Hosting/ResourceBuilderExtensions.cs

```diff
@@ -1139,6 +1140,289 @@ public static IResourceBuilder<T> WithCommand<T>(
+    /// </para>
+    /// <para>
+    /// The <paramref name="configureRequest"/> callback will be invoked to configure the request before it is sent. This can be used to add headers or a request payload
+    /// before the request is sent.
+    /// </para>
+    /// <para>
+    /// The <paramref name="getCommandResult"/> callback will be invoked after the response is received to determine the result of the command invocation. If this callback
+    /// is not specified, the command will be considered succesful if the response status code is in the 2xx range.
+    /// </para>
+    /// </remarks>
+    public static IResourceBuilder<TResource> WithHttpCommand<TResource>(
+        this IResourceBuilder<TResource> builder,
+        string path,
+        string displayName,
+        Func<EndpointReference>? endpointSelector,
+        HttpMethod? method = null,
+        string? httpClientName = null,
+        Func<HttpRequestMessage, Task>? configureRequest = null,
+        Func<HttpResponseMessage, Task<ExecuteCommandResult>>? getCommandResult = null,
+        string? commandName = null,
+        string? displayDescription = null,
+        string? confirmationMessage = null,
+        string? iconName = null,
+        IconVariant? iconVariant = null,
+        bool isHighlighted = false)
+        where TResource : IResourceWithEndpoints
+    {
+        method ??= HttpMethod.Post;
+
+        endpointSelector ??= PreferredEndpointSelector(builder);
+
+        var endpoint = endpointSelector()
+            ?? throw new DistributedApplicationException($"Could not create HTTP command for resource '{builder.Resource.Name}' as the endpoint selector returned null.");
+
+        commandName ??= $"{endpoint.Resource.Name}-{endpoint.EndpointName}-http-{method.Method.ToLowerInvariant()}";
+
+        var targetRunning = false;
+        builder.ApplicationBuilder.Eventing.Subscribe<BeforeStartEvent>((e, ct) =>
+        {
+            var rns = e.Services.GetRequiredService<ResourceNotificationService>();
+            var watchTask = Task.Run(async () =>
+            {
+                await foreach (var resourceEvent in rns.WatchAsync(ct).WithCancellation(ct))
+                {
+                    if (resourceEvent.Resource == endpoint.Resource)
+                    {
+                        var resourceState = resourceEvent.Snapshot.State?.Text;
+                        targetRunning = resourceState == KnownResourceStates.Running || resourceState == KnownResourceStates.RuntimeUnhealthy;
+                    }
+                }
+            }, ct);
+
+            return Task.CompletedTask;
+        });
+
+        builder.WithCommand(commandName, displayName,
+            async context =>
+            {
+                if (!endpoint.IsAllocated)
+                {
+                    return new ExecuteCommandResult { Success = false, ErrorMessage = "Endpoints are not yet allocated." };
+                }
```

Comment:
Does it make sense to instead not enable the command if the endpoint isn't allocated?

Reply:
It does that already (see the `updateState` arg below), but this is the actual execute method so it's too late to disable it. If we somehow hit this case, this code means we'll just return a failed command result.

Reply:
There could be a delay between state being updated, the user clicking the button, and the request hitting the server. Server validation of state is still good.